### PR TITLE
close txn rightly

### PR DESCRIPTION
**Pull request by daviszhen**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [x] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #

## What this PR does / why we need it:

replace all goto by the defer to finish the transaction rightly.

I have tested it on the branch https://github.com/daviszhen/matrixone/tree/0527-defer-txn five times with 30 seconds + panic enabled. No further panic has been found.

### CODE REVIEW

**Comment by matrix-meow** [comment]
@daviszhen Thanks for your contributions!

**Here are review comments for file pkg/frontend/authenticate.go:**
This pull request lacks sufficient information and context to be properly reviewed. The title is vague and does not provide any indication of what specific change is being made. The body of the pull request is also lacking in detail, with no explanation of why this change is necessary or what problem it solves. Additionally, the changes made in the pull request are empty, which makes it impossible to review the actual code changes. 

To improve this pull request, the title should be updated to provide a clear and concise summary of the change being made. The body should include a detailed explanation of the problem being addressed and how this change solves it. The changes made in the pull request should also be included so that reviewers can see the actual code changes and provide feedback on them. 

Furthermore, it would be helpful to include any relevant documentation or test cases to ensure that the change is properly tested and documented. Finally, it is important to ensure that the code changes are optimized and do not introduce any new issues or bugs.

**Here are review comments for file pkg/frontend/authenticate_test.go:**
The title of the pull request is not descriptive enough. It only says "change txn" which does not provide any information about what the pull request is trying to achieve. The body of the pull request is also lacking in detail. It only mentions that all "goto" statements have been replaced with "defer" statements to finish the transaction correctly. However, it does not explain why this change is necessary or how it will improve the codebase.

The changes made in the pull request are only for one file, "pkg/frontend/authenticate_test.go". It adds a new function "newMrsForSystemVariablesOfAccount" and modifies the existing function "newMrsForPrivilegeWGO". However, there is no explanation of why these changes were made or how they relate to the overall goal of the pull request.

To improve the pull request, the title should be more descriptive, such as "Replace all goto statements with defer statements to improve transaction handling". The body should provide more context on why this change is necessary and how it will improve the codebase. Additionally, the changes made in the file should be explained in more detail, including why the new function was added and how it relates to the overall goal of the pull request. Finally, the pull request should include any relevant tests to ensure that the changes do not introduce any new bugs or issues.

**Here are review comments for file pkg/frontend/compiler_context.go:**
The title and body of the pull request are concise and clear. However, there are a few issues with the changes made in the pull request that need to be addressed:

1. The changes made to the `ResolveUdf` function are not optimal. The use of `goto` statements has been replaced with `defer` statements to finish the transaction rightly. However, the code still contains a lot of error handling using `if` statements, which makes the code harder to read and understand. It would be better to use a single `if err != nil` statement at the end of the function to handle all errors.

2. The `ResolveAccountIds` function also needs to be optimized. The use of `goto` statements has been replaced with `defer` statements to finish the transaction rightly. However, the code still contains a lot of error handling using `if` statements, which makes the code harder to read and understand. It would be better to use a single `if err != nil` statement at the end of the function to handle all errors.

3. The `ResolveUdf` function has a redundant variable declaration for `body`. It is declared twice, once at the beginning of the function and once inside the `if execResultArrayHasData(erArray)` block. This can be simplified by removing the first declaration.

To optimize the changes made in the pull request, the following suggestions can be made:

1. Use a single `if err != nil` statement at the end of the `ResolveUdf` and `ResolveAccountIds` functions to handle all errors.

2. Remove the redundant variable declaration for `body` in the `ResolveUdf` function.

3. Consider using a more descriptive variable name than `body` in the `ResolveUdf` function to make the code easier to understand.

4. Consider adding comments to the code to explain the purpose of each function and variable. This will make the code easier to understand for other developers who may need to work on it in the future.

**Here are review comments for file pkg/frontend/mysql_cmd_executor_test.go:**
The pull request title is not descriptive enough. It only says "change txn" without providing any context or information about what the changes are. This can make it difficult for other developers to understand the purpose of the pull request. A more descriptive title would be helpful.

The pull request body is brief and lacks detail. It only provides a checklist of the type of PR and the reason for the changes. It would be helpful to provide more information about what the changes are, why they are necessary, and how they will impact the codebase.

The changes made in the pull request are not optimized. The changes only replace a single line of code that uses a goto statement with a defer statement. While this change may improve the readability of the code, it does not provide any significant optimization or performance improvement.

There are no major issues with the changes made in the pull request. However, it would be helpful to provide more context and detail about the changes in the pull request body. Additionally, the title should be more descriptive to help other developers understand the purpose of the changes.

**Here are review comments for file pkg/frontend/plsql_interpreter.go:**
The title of the pull request is not descriptive enough. It only mentions "change txn" which is not clear enough to understand what the pull request is about. It would be better to provide a more detailed and descriptive title that explains the changes made in the pull request.

The body of the pull request is also not informative enough. It only provides a brief description of the changes made and does not explain why these changes are necessary. It would be better to provide more context and details about the changes made and why they are important.

In terms of the changes made, the pull request replaces all "goto" statements with "defer" statements to finish the transaction correctly. However, there are some issues with the changes made:

1. The "handleFailed" label is no longer used, but the code still refers to it. This can cause confusion and should be removed.

2. The "err" variable is being overwritten in the "defer" statement. This can cause unexpected behavior and should be avoided.

To optimize the changes made, it would be better to use a single "defer" statement to handle both the commit and rollback operations. This would simplify the code and make it easier to read and maintain. Additionally, it would be better to provide more detailed comments explaining the purpose of each section of the code. This would make it easier for other developers to understand the code and make changes if necessary.

**Here are review comments for file pkg/frontend/session.go:**
The title and body of the pull request are clear and concise. However, there are several issues with the changes made in the pull request:

1. The commit message is not descriptive enough. The commit message should provide more information about the changes made in the pull request.

2. The changes made in the pull request are not optimized. The changes made in the pull request replace all `goto` statements with `defer` statements to finish the transaction rightly. However, this change may not be the most efficient way to handle transactions. It would be better to use a transaction manager or a database library that handles transactions automatically.

3. The variable name `isNotBackgroundSession` is not clear. It would be better to use a more descriptive variable name, such as `isForegroundSession`.

4. The `ExecSql` function has a lot of duplicated code. It would be better to refactor the code to reduce duplication.

To address these issues, the following changes could be made:

1. The commit message should be updated to provide more information about the changes made in the pull request.

2. Instead of replacing all `goto` statements with `defer` statements, it would be better to use a transaction manager or a database library that handles transactions automatically.

3. The variable name `isNotBackgroundSession` should be changed to `isForegroundSession`.

4. The `ExecSql` function should be refactored to reduce duplication. One way to do this would be to extract the transaction handling code into a separate function.

**Here are review comments for file pkg/frontend/show_account.go:**
The title of the pull request is not descriptive enough. It simply states "change txn" without providing any context or explanation of what the change entails. This makes it difficult for reviewers to understand the purpose of the pull request and provide meaningful feedback.

The body of the pull request is also lacking in detail. It only includes a checklist of PR types and a brief statement about replacing all "goto" statements with "defer" to finish the transaction correctly. This statement does not provide enough information about why this change is necessary or how it will improve the codebase.

In terms of the changes made in the pull request, there are a few issues that need addressing. Firstly, the "doShowAccounts" function has been modified to return an error instead of using "goto" statements. However, the error handling logic is not consistent throughout the function. In some cases, errors are returned immediately, while in others, they are assigned to the "err" variable and then handled later. This inconsistency could lead to unexpected behavior and should be addressed.

Additionally, the "begin" and "commit" statements have been moved to a defer function, which is a good improvement. However, the "rollback" statement is still being called in the "handleFailed" label, which is not necessary since the defer function will handle rolling back the transaction if an error occurs.

To optimize the changes made in the pull request, it would be helpful to include more detailed comments in the code to explain the purpose of each change. This would make it easier for reviewers to understand the changes and provide feedback. Additionally, it would be beneficial to include test cases to ensure that the changes do not introduce any new bugs or issues.

**Here are review comments for file pkg/frontend/show_publications.go:**
The title of the pull request is not descriptive enough. It only says "change txn" which does not provide any information about what the pull request is trying to achieve. It would be better to have a more descriptive title that summarizes the changes made in the pull request.

In the body of the pull request, the issue number that the pull request is fixing is missing. It is important to include the issue number so that it is easier to track the changes made in the pull request.

The changes made in the pull request are replacing all `goto` statements with `defer` statements to finish the transaction correctly. However, the code changes made in the pull request are not optimized. The `err` variable is being shadowed in the `defer` statement which can cause confusion and make the code harder to read. It would be better to use a different variable name to avoid shadowing the `err` variable.

Additionally, the `goto` statements are being replaced with `return` statements which is a better way to handle errors. However, the `handleFailed` label is still present in the code which is not needed anymore. It would be better to remove the `handleFailed` label and the code associated with it.

Finally, the `commit` statement is being executed outside the `defer` statement which can cause issues if there is an error after the `commit` statement is executed. It would be better to move the `commit` statement inside the `defer` statement to ensure that it is executed even if there is an error.

**Here are review comments for file pkg/frontend/show_subsriptions.go:**
The title of the pull request is not descriptive enough. It only says "change txn" which does not give any information about what the pull request is trying to achieve. A more descriptive title would be helpful.

The body of the pull request is also not informative enough. It only mentions that the pull request replaces all goto statements with defer statements to finish the transaction correctly. It does not explain why this change is necessary or what benefits it brings. A more detailed explanation would be helpful.

In the changes made to the file `pkg/frontend/show_subsriptions.go`, the function `doShowSubscriptions` has been modified to use a `defer` statement to finish the transaction correctly. However, the `goto` statements have not been replaced completely. Instead, they have been replaced with `return` statements. This is not a good practice as it makes the code harder to read and understand. It would be better to replace the `goto` statements with `return` statements in a more structured way.

Additionally, the `handleFailed` label is not being used anymore. It would be better to remove it completely.

Finally, the `commit` statement has been removed from the function. It is unclear why this has been done. If the transaction is not committed, the changes made to the database will not be saved. It would be better to keep the `commit` statement in the function.

**Here are review comments for file pkg/frontend/util.go:**
This pull request seems to be addressing an improvement in the codebase. However, there are several issues with the title, body, and changes that need to be addressed before merging the pull request.

1. Title: The title "change txn" is not descriptive enough and does not provide any context about the changes made in the pull request. A more descriptive title would be helpful for reviewers to understand the purpose of the pull request.

Suggestion: Change the title to "Replace all goto statements with defer to properly finish transactions"

2. Body: The body of the pull request is incomplete and does not provide enough information about the changes made. It only mentions that all goto statements have been replaced with defer to finish transactions properly. It would be helpful to provide more context about why this change was necessary and how it improves the codebase.

Suggestion: Add more information to the body of the pull request, such as why goto statements were problematic and how using defer improves the codebase.

3. Changes: The changes made in the pull request include adding a new function to util.go, which is not related to the main purpose of the pull request. This change should be separated into a different pull request to keep the changes focused on the main purpose.

Suggestion: Remove the addition of the new function from the pull request and create a separate pull request for that change.

4. Optimization: While the change from goto to defer is an improvement, there may be other ways to optimize the code further. It would be helpful to review the codebase and identify any other areas that could be improved.

Suggestion: Review the codebase and identify any other areas that could be improved, such as reducing code duplication or improving performance.

**Review by qingxinhome — APPROVED** [review]

File: pkg/frontend/compiler_context.go
Comment (fengttt) [inline pkg/frontend/compiler_context.go:293]: here the err as params to finishTxn, is the err from bh.Exec(ctx, "begin;").   I think you need the last err before return.    Is this right?

If you think I am wrong, please find some go expert to review this and confirm.

**Review by fengttt — CHANGES_REQUESTED** [review]

File: pkg/frontend/compiler_context.go
Comment (daviszhen) [reply pkg/frontend/compiler_context.go:293]: the `err` is defined at the `func (tcc *TxnCompilerContext) ResolveUdf(name string, args []*plan.Expr) (body string, err error)`. it is shared among the function.  it is returned the last err always.

**Review by fengttt — APPROVED** [review]

**Review by fengttt — APPROVED** [review]

**Review by fengttt — APPROVED** [review]

**Review by fengttt — APPROVED** [review]

**Review by fengttt — APPROVED** [review]

**Pull request merged**
